external/fv3fit/tests/training/test_main.py

```diff
@@ -86,86 +105,135 @@ def mock_load_batches():
+    return CallArtifacts(
+        config.output_path,
+        all_variables,
+        MockDerivedModel,
+        config.input_variables,
+        config.output_variables,
+        hyperparameters,
+    )
+
+
 @pytest.mark.parametrize("additional_variables", [[], ["downward_shortwave"]])
-def test_main(
-    mock_dataset: xr.Dataset,
+@pytest.mark.parametrize("derived_output_variables", [[], ["downwelling_shortwave"]])
+def test_main_calls_load_batches_correctly(
     tmpdir,
     mock_load_batches: mock.MagicMock,
     mock_train_dense_model: mock.MagicMock,
     additional_variables: Sequence[str],
+    derived_output_variables: Sequence[str],
 ):
     """
     Test of fv3fit.train main function only, using mocks for training function
     and data loading.
     """
-    mock_load_batches.return_value = [mock_dataset for _ in range(6)]
-    args, input_variables, output_variables, hyperparameters, output_path = get_config(
-        tmpdir, mock_dataset, additional_variables, []
+    artifacts = call_main(
+        tmpdir, mock_load_batches, additional_variables, derived_output_variables,
     )
-    all_variables = input_variables + output_variables + additional_variables
-    fv3fit.train.main(args)
-    mock_load_batches.assert_called_with(variables=all_variables)
+    mock_load_batches.assert_called_with(variables=artifacts.all_variables)
     assert mock_load_batches.call_args_list[0] == mock_load_batches.call_args_list[1]
     assert mock_load_batches.call_count == 2
-    mock_train_dense_model.assert_called_once_with(
-        input_variables=input_variables,
-        output_variables=output_variables,
-        hyperparameters=hyperparameters,
-        train_batches=mock_load_batches.return_value,
-        validation_batches=mock_load_batches.return_value,
+
+
+@pytest.mark.parametrize("additional_variables", [[], ["downward_shortwave"]])
+@pytest.mark.parametrize("derived_output_variables", [[], ["downwelling_shortwave"]])
+def test_main_dumps_correct_predictor(
+    tmpdir,
+    mock_load_batches: mock.MagicMock,
+    mock_train_dense_model: mock.MagicMock,
+    additional_variables: Sequence[str],
+    derived_output_variables: Sequence[str],
+):
+    """
+    Test of fv3fit.train main function only, using mocks for training function
+    and data loading.
+    """
+    artifacts = call_main(
+        tmpdir, mock_load_batches, additional_variables, derived_output_variables,
     )
     mock_predictor = mock_train_dense_model.return_value
-    mock_predictor.dump.assert_called_once()
-    assert output_path in mock_predictor.dump.call_args[0]
+    if len(derived_output_variables) > 0:
+        dump_predictor = artifacts.MockDerivedModel.return_value
+    else:
+        dump_predictor = mock_predictor
+    dump_predictor.dump.assert_called_once()
+    assert artifacts.output_path in dump_predictor.dump.call_args[0]
 
 
 @pytest.mark.parametrize("additional_variables", [[], ["downward_shortwave"]])
```

Comment:
Same comment as above re: `additional_variables`

Reply:
It's possible for a bug to make this matter.

In each of these tests, I would maybe exclude the less-likely-to-matter cases if they took >1s to run, but all the main tests collectively take much less than a second to run because of the mocks.